### PR TITLE
Group handling & groupMessage handling (unfinished)

### LINKED ISSUE #25: Users are not notified about message status changes during handshake
**Describe the bug**
When a user logs in and performs a handshake, they receive any unread messages left for them on the server. This does not apply to `MessageStatusChangeEvent`s that are processed while the user is offline.

**To Reproduce**
1. Start the server and a client
2. Send a message to another client
3. Disconnect the client
4. Start the other client
5. Read the message
6. Start the first client again

**Expected behavior**
The message is displayed with the current status (either `RECEIVED` or `READ`).

**Additional context**
This can probably be solved with an additional query being executed during the handshake and an additional `Cache<MessageStatusChangeEvent>` on the client to process these updates.

### PR DESCRIPTION
**Pull request by DieGurke**
* Group Handling
* Group Message handling
* Group Message Status handling (unfinished)
* Message Status now works properly (RECEIVED)

Closes #25

### CODE REVIEW

File: src/main/java/envoy/server/data/GroupMessage.java [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+
+/**
+ * Project: <strong>envoy-server-standalone</strong><br>
+ * File: <strong>GroupMessage.java</strong><br>
+ * Created: <strong>18.04.2020</strong><br>
+ * 
+ * @author Maximilian K&auml;fer
+ * @since Envoy Server Standalone v0.1-beta
+ */
+
```
Comment (CyB3RC0nN0R) [inline src/main/java/envoy/server/data/GroupMessage.java:21]:
```suggestion
```

File: src/main/java/envoy/server/data/Message.java [hunk 2]
```diff
@@ -2,7 +2,16 @@
...
+import javax.persistence.Entity;
+import javax.persistence.Id;
+import javax.persistence.Inheritance;
+import javax.persistence.InheritanceType;
+import javax.persistence.JoinColumn;
+import javax.persistence.ManyToOne;
+import javax.persistence.NamedQuery;
+import javax.persistence.Table;
+import javax.persistence.Temporal;
+import javax.persistence.TemporalType;
```
Comment (CyB3RC0nN0R) [inline src/main/java/envoy/server/data/Message.java:14]:
```suggestion
import javax.persistence.*;
```

File: src/main/java/envoy/server/net/ConnectionManager.java [hunk 3]
```diff
@@ -44,60 +46,74 @@ private ConnectionManager() {}
...
+	 *
+	 * @param group the group to search for
+	 * @return a set of all IDs of currently active members in this group
+	 * @since Envoy Server Standalone v0.1-beta
+	 */
+	public Set<Long> getOnlineUsersOfGroup(Group group) {
+		Set<Long>	onlineMembers	= new HashSet<>();
+		Set<Long>	members			= group.getContacts().stream().map(envoy.server.data.Contact::getID).collect(Collectors.toSet());
+		members.forEach(userID -> { if (isOnline(userID)) onlineMembers.add(userID); });
+		return onlineMembers;
```
Comment (CyB3RC0nN0R) [inline src/main/java/envoy/server/net/ConnectionManager.java:117]: Please test this before merging!
```suggestion
		return group.getContacts().stream().map(envoy.server.data.Contact::getID).filter(this::isOnline).collect(Collectors.toSet());
```

File: src/main/java/envoy/server/processors/GroupResizeProcessor.java [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+		// Send the updated group to all of its members
+		var commonGroup = group.toCommon();
+		group.getContacts()
+			.stream()
+			.map(Contact::getID)
+			.filter(connectionManager::isOnline)
+			.map(connectionManager::getSocketId)
+			.forEach(memberSocketID -> {
+				try {
+						writeProxy.write(memberSocketID, commonGroup);
```
Comment (CyB3RC0nN0R) [inline src/main/java/envoy/server/processors/GroupResizeProcessor.java:52]:
```suggestion
					writeProxy.write(memberSocketID, commonGroup);
```

File: src/main/java/envoy/server/processors/GroupMessageProcessor.java [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+public class GroupMessageProcessor implements ObjectProcessor<GroupMessage> {
+
+	private static final Logger logger = EnvoyLog.getLogger(GroupCreationProcessor.class);
+
+	@Override
+	public void process(GroupMessage groupMessage, long socketID, ObjectWriteProxy writeProxy) {
+		groupMessage.nextStatus();
+		ConnectionManager connectionManager = ConnectionManager.getInstance();
+		try {
+			writeProxy.write(socketID, new MessageStatusChangeEvent(groupMessage));
```
Comment (CyB3RC0nN0R) [inline src/main/java/envoy/server/processors/GroupMessageProcessor.java:34]: Why do we send the status change back to the sender of the message? I thought that the sender increments the status of the message on their own.

File: src/main/java/envoy/server/processors/GroupMessageProcessor.java [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+
+		final var members = PersistenceManager.getInstance().getGroupByID(groupMessage.getRecipientID()).getContacts();
+		for (long i = 0; i < groupMessage.getMemberStatuses().size(); i++) {
+			groupMessage.getMemberStatuses().replace(i, MessageStatus.SENT);
+		}
+		members.forEach(user -> { setMemberStatus(connectionManager, groupMessage, user.getID()); });
+
+		// Checks if all memberMessageStatuses are RECEIVED and if so sets the
+		// groupMessage Status to RECEIVED.
+		if (!groupMessage.getMemberStatuses().containsValue(MessageStatus.SENT)) groupMessage.setStatus(MessageStatus.RECEIVED);
```
Comment (delvh) [inline src/main/java/envoy/server/processors/GroupMessageProcessor.java:41]:
```suggestion		groupMessage.setStatus(Collections.min(groupMessage.getMemberStatuses()))
```

File: src/main/java/envoy/server/net/ConnectionManager.java [hunk 7]
```diff
@@ -44,60 +50,71 @@ private ConnectionManager() {}
...
 	}
 
 	/**
-	 * @param userId the ID of the user registered at a socket
+	 * @param userID the ID of the user registered at a socket
 	 * @return the ID of the socket
 	 * @since Envoy Server Standalone v0.1-alpha
 	 */
-	public long getSocketId(long userId) { return sockets.get(userId); }
+	public long getSocketId(long userID) { return sockets.get(userID); }
```
Comment (delvh) [inline src/main/java/envoy/server/net/ConnectionManager.java:86]: - [ ] Please refactor to `getSocketID`

File: src/main/java/envoy/server/processors/MessageProcessor.java [hunk 8]
```diff
@@ -17,28 +22,46 @@
...
+		sendToUser(connectionManager, message, writeProxy);
+		if (message.getStatus() != MessageStatus.SENT) {
+			// Sending a messageStatusChangeEvent to the sender
+			try {
+				writeProxy.write(socketID, new MessageStatusChangeEvent(message));
+			} catch (IOException e) {
+				logger.warning("Could not send messageStatusChangeEvent to the sender of this message with ID: " + message.getID());
+				e.printStackTrace();
+			}
+		}
```
Comment (delvh) [inline src/main/java/envoy/server/processors/MessageProcessor.java:46]:
```suggestion
		if (message.getStatus!=MessageStatus.WAITING) {
			logger.warning("Received message with invalid status: " + message);
			return;
		}
		message.nextStatus();
		ConnectionManager	connectionManager	= ConnectionManager.getInstance();

		sendToUser(connectionManager, message, writeProxy);
```

File: src/main/java/envoy/server/processors/MessageStatusChangeProcessor.java [hunk 9]
```diff
@@ -23,17 +23,17 @@
...
-		envoy.server.data.Message msg = persistenceManager.getMessageById(input.getID());
+		envoy.server.data.Message msg = persistenceManager.getMessageByID(input.getID());
 		msg.setStatus(input.get());
 		msg.setReadDate(input.getDate());
 		persistenceManager.updateMessage(msg);
 
 		// Notifies the sender of the message about the status-update to READ
-		final long senderId = msg.getSender().getId();
+		final long senderId = msg.getSender().getID();
 		if (connectionManager.isOnline(senderId)) writeProxy.write(connectionManager.getSocketId(senderId), input);
```
Comment (delvh) [inline src/main/java/envoy/server/processors/MessageStatusChangeProcessor.java:37]:
```suggestion
		final long senderID = msg.getSender().getID();
		if (connectionManager.isOnline(senderID)) writeProxy.write(connectionManager.getSocketId(senderID), input);
```

File: src/main/resources/META-INF/persistence.xml [hunk 10]
```diff
@@ -3,25 +3,25 @@
...
-				value="org.hibernate.dialect.PostgreSQL95Dialect" /> <!-- DB Dialect -->
-			<property name="hibernate.hbm2ddl.auto" value="update" /> <!-- create / create-drop / update -->
+				value="org.hibernate.dialect.PostgreSQL95Dialect" />
+			<property name="hibernate.hbm2ddl.auto" value="update" />
 
 		</properties>
-
 	</persistence-unit>
-
 </persistence>
```
Comment (delvh) [inline src/main/resources/META-INF/persistence.xml:27]:
```suggestion
</persistence>

```

**Pull request merged**